### ISSUE TITLE
Atomic saving

### ISSUE THREAD

**Issue by lazka** [quodlibet/mutagen#241]
Originally reported by: **Daniel Plachotich (Bitbucket: [danpla](http://bitbucket.org/danpla), GitHub: [danpla](http://github.com/danpla))**

---

Resizing files "in place" is pretty dangerous way, especially for the large ones like audio. Both memory mapping and ordinary IO can fail during the process for many reasons and leave a corrupt file.

The only safe approach is to use temporary file. This is already used in some players and tagging libraries, for example, in Foobar2000 [[1](https://www.hydrogenaud.io/forums/index.php?s=21f784e988b9216cb9f6c9e4206194e6&showtopic=110298&view=findpost&p=909095)][[2](https://www.hydrogenaud.io/forums/index.php?showtopic=99822&st=0&p=826803&#entry826803)][[3](http://foosion.foobar2000.org/doxygen/0.9.4/classtag__write__callback.html)].

The algorithm is pretty simple:
- Rename the original file (for example, by appending ".orig" to the end)
- Create a new one with the name of original and write to it:
  - New tags
  - Copy data from the original file
- On success, remove the original

---
- Bitbucket: https://bitbucket.org/lazka/mutagen/issue/241

**Comment by Lanchon** [comment]
FYI:

this link discusses the create/write/fsync/rename idiom:
http://stackoverflow.com/questions/2333872/atomic-writing-to-file-with-python

regarding this code:

```
f = open(tmpFile, 'w')
f.write(text)
# make sure that all data is on disk
# see http://stackoverflow.com/questions/7433057/is-rename-without-fsync-safe
f.flush()
os.fsync(f.fileno()) 
f.close()

os.rename(tmpFile, myFile)
```

it says:

> If successful, the renaming will be an atomic operation (this is a POSIX requirement). On Windows, if dst already exists, OSError will be raised even if it is a file; there may be no way to implement an atomic rename when dst names an existing file

also see this old project:
https://github.com/sashka/atomicfile
https://github.com/sashka/atomicfile/search?q=windows&type=Issues

and this notes from microsoft:
https://msdn.microsoft.com/en-us/library/windows/desktop/aa365241%28v=vs.85%29.aspx
https://msdn.microsoft.com/en-us/library/windows/desktop/hh802690(v=vs.85).aspx#applications_updating_a_single_file_with_document-like_data

**Comment by lazka** [comment]
Thanks.

I've implemented something similar in quodlibet: https://github.com/quodlibet/quodlibet/blob/master/quodlibet/quodlibet/util/atomic.py

**Comment by lazka** [comment]
Proposed API:

```
def replace(self, filename=None):
    """Replaces the file with a copy including all tag changes. In case it
    fails or gets interrupted the file will be unchanged. If it returns
    successfully the file will be replaced and all file system changes will be
    flushed to disk.

    This will prevent corrupt files due to bugs in mutagen or if the system
    stops working due to a power failure, crashes etc.

    This operation can fail in cases where save() would succeed.
    This operation can be significantly slower than save().

    Args:
        filename (fspath): The file to save to or None to use the file
            loaded from
    Raises:
        MutagenError: in case of an error
    """

    raise NotImplementedError
```

**Comment by Lanchon** [comment]
> Replaces the file with a copy

> all file system changes will be flushed to disk.

these are implementation details and shouldn't be in the API. atomicity is the key property here. implementation strategy and durability should not be specified.

> def replace(self, filename=None)

IMHO this is the wrong API. no file system in real use that i know of is fully transactional. this API provides and abstraction that is not well supported underneath. in the end, it will create a copy of the file in all supported OSes; this is why i think it is the wrong API.

a more useful API would be saveAs(self, file=None) that saves pending changes to a copy of the file somewhere else. API help would include a note that, if 'file' parameter is missing, it will save to a temp file and then replace the original as atomically as possible. (and stating that this saving mode is safer and slower than the usual save is OK; but it should be noted that the usual save is safe enough for normal use.)

> This operation can be significantly slower than save().

yes, the saveAs() API could be added. but the important issue here is that it doesn't fix the problem with the current implementation of save()! furthermore, you leave it to the app programmer to decide what to do, and they know little or nothing about risks. what do you expect them to do? wouldn't all of them either choose replace()/saveAs() or ask you a question personally?

plus the 'safe save' is only needed when the ID3v2 tag is added or resized, because loosing the tag is not as bad a loosing the payload, and the tag is so small that a partial save is really unlikely. and with replace/saveAs you force a copy on each save, when 99% of the time it is not needed.

what is really needed is a safe 'payload scroll' only when resizing the ID3v2 tag. IMO, the normal save should be upgraded to behave in a **safe enough** manner, and that is it. this entails copying the file but only when a payload move is needed. more free space is needed by the new save? too bad. i rather have a reasonably safe save that complains the disk is full than a dangerous save. if the disk is full, users will have to deal with that very soon, so let them deal with it now.

so to recap, my proposal is:

1. fix save(): its current implementation is unsafe and unfit for all uses. most existing client programs will never upgrade to replace() anyway, so replace won't fix the problem. most newer clients will choose replace() and your library will become slow and hated.

2. optionally (and at a later time) add saveAs() API. include with it a proper implementation of "save to temp and rename" when no target file is given.

**Comment by Lanchon** [comment]
and comparatively the save/replace solution has problems:

1. save() is unfit for all uses given that it is unsafe. it should be fixed or removed from the API.

2. replace() is unfit for most uses given that its implementation is terribly inefficient most of the time.

keep in mind that the shortcoming of current save implementation is not a 'designed feature'. it is not a calculated compromise of performance and safety; it is a plain-old bug in the implementation, arising from not considering certain failure modes. its behavior is not something anyone would want to keep, but a bug that should be fixed.

fortunately the save() API is the right API for the fixed implementation (a safe enough save), so you can fix this without impacting clients.

**Comment by lazka** [comment]
Thanks for your input.

> > Replaces the file with a copy
> > all file system changes will be flushed to disk.
>
> these are implementation details and shouldn't be in the API. atomicity is the key property here. implementation strategy and durability should not be specified.

I didn't want to be too specific as in reality the drive/file system can lie about flushing and on Windows we don't know what's happening exactly. We could say it does an atomic replace if the OS/hardware supports it, if not it will at least reduce the probability of data corruption as far as possible (?)

> > def replace(self, filename=None)
>
> IMHO this is the wrong API. no file system in real use that i know of is fully transactional. this API provides and abstraction that is not well supported underneath. in the end, it will create a copy of the file in all supported OSes; this is why i think it is the wrong API.

Why is it wrong?

> a more useful API would be saveAs(self, file=None) that saves pending changes to a copy of the file somewhere else. API help would include a note that, if 'file' parameter is missing, it will save to a temp file and then replace the original as atomically as possible. (and stating that this saving mode is safer and slower than the usual save is OK; but it should be noted that the usual save is safe enough for normal use.)

There might be a misunderstanding here. I planed exactly what you propose with replace().
The filename argument allows to save tags to another file than loaded from, like``ID3().replace("foo.mp3")``. Imo temp file creation should be an implementation detail as the only logical place to put it in is the same directory. Or can you think of a use case providing your own temp file?

> > This operation can be significantly slower than save().
> 
> yes, the saveAs() API could be added. but the important issue here is that it doesn't fix the problem with the current implementation of save()! furthermore, you leave it to the app programmer to decide what to do, and they know little or nothing about risks. what do you expect them to do? wouldn't all of them either choose replace()/saveAs() or ask you a question personally?

Good question. I'll have to think about that.

> plus the 'safe save' is only needed when the ID3v2 tag is added or resized, because loosing the tag is not as bad a loosing the payload, and the tag is so small that a partial save is really unlikely. and with replace/saveAs you force a copy on each save, when 99% of the time it is not needed.

I don't like multiple shades of "safe". If we somehow leave the file in an invalid but playable state, the next tagger might get tricked into corrupting the file, or some players wont accept the file. And I really don't want to think at every operation in the code about what would happen if there was a crash right after. And is there any guarantee that file operations get flushed to disk in the order I execute them?

> what is really needed is a safe 'payload scroll' only when resizing the ID3v2 tag. IMO, the normal save should be upgraded to behave in a **safe enough** manner, and that is it. this entails copying the file but only when a payload move is needed. more free space is needed by the new save? too bad. i rather have a reasonably safe save that complains the disk is full than a dangerous save. if the disk is full, users will have to deal with that very soon, so let them deal with it now.

I'm not sure, if we talk probability here, how often the "safe enough" approach would result in a recoverable file (from the user POV, not starting up the HEX editor and cleaning things up by hand). (Don't forget all the other tagging formats..).

I agree that free space, permissions, and path limits reached by the temp file should just be ignored/error out.

> so to recap, my proposal is:
> 
> 1. fix save(): its current implementation is unsafe and unfit for all uses. most existing client programs will never upgrade to replace() anyway, so replace won't fix the problem. most newer clients will choose replace() and your library will become slow and hated.

I think you are exaggerating a bit. What amount of programs do atomic replace for your documents? If your kernel crashes a lot you have lots of other problems too. I at least hope that kernel bugs like the one you reported are rare.

> 2. optionally (and at a later time) add saveAs() API. include with it a proper implementation of "save to temp and rename" when no target file is given.

Regarding the naming. I prefer replace() as it highlights the the file gets replaced as a whole, while when I read saveAs() I think of saving to another location (like the "save as" menu entry in many file menus), and while that is a step of the process, it's just an implementation detail.

**Comment by Lanchon** [comment]
well, what can i say, my view is different.

IMHO save() is broken because it causes a very high probability of loosing data, compared to any other saves the users normally do. to worsen things, the data lost is not the data being edited (the audio), and this is the worst scenario and something users won't understand or be lenient with. save() has to be fixed because it is broken and has not use case as it is. there is no point in introducing another API and leaving a broken one up. and expecting existing clients to change is not realistic.

replace() is not the solution, as it causes massive performance loss all the time, when most of the time it is not needed.

the probability that a 1KB structure gets partially written is abysmally low compared to the probability that a 100MB file gets partially written; maybe 100.000 times smaller. (i listen to dj sets mostly, so my individual files are typically over 100MB.) this means that around 100.000 corruption incidents of moved tracks would be observed before 1 tag-only corruption happens (if both operations are equally likely to be executed). but of course, before 100.000 corruptions happen your library would be eliminated from any successful app.

i guess 100.000 times safer than the current implementation can be called 'safe enough', given that you are not currently bombarded by thousands of corruption reports.

> Regarding the naming. I prefer replace() as it highlights the the file gets replaced as a whole, while when I read saveAs() I think of saving to another location

this is exactly what saveAs() does: saves a new file. i proposed that omitting the argument causes save as temp file and replacement of the original. ok, maybe this is not the best interface. so instead: saveAs() must be called with a file argument, and the file must not exist. saveAtomically(), saveViaCopy(), atomicSave(), saveSafely(), safeSave() are all possible names of "saveAs() to temp and then atomically replace". the docs should state that although atomicSave() is safer, save() should be is safe enough for all uses. (implementation of atomicSave() has a low priority IMHO because it will almost never be used. fixing save() has a high priority.)

replace() is a bad name, and the semantics are not at all clear. why would you call replace("xxx.mp3") to save a new xxx.mp3 file? and if you want to save to a new file, why would the provided API silently replace any existing file instead of complaining that it exists? and most importantly: if the file exists, would replace() replace it atomically? if not, how would you document that? if yes, then there is a 3rd filename involved that is not being specified!

my semantics:
save() is save in-place. reasonably safe and reasonably performant. (creates a temp copy only if a dangerous operation is needed.)
saveAs() is save to new file. no ugly replacement semantics: the method just returns an error if the file exists.
saveAtomically() is an ultra-safe, ultra slow implementation of save in-place. included for completeness and because it is easy to implement once you have saveAs().

> What amount of programs do atomic replace for your documents?

all of them i suppose. the problem is that some apps didn't do an fsync so data losses happened all the same when people changed file systems. you can read about it: https://lwn.net/Articles/322823/

the point is that data loss happens a lot. if your library is used that is; if nobody uses it, then yes, whatever you do here is just fine :). but the idea is that a library must be robust **just in case** it becomes successful. this is why IMHO save() has to be fixed.

i know that adding code to implement a decision to copy or not before starting the save process to the existing code base can be painful. but you don't need to do that. you can start saving in-place and only create a temp when you need to scroll big data. after the scroll, you immediately replace atomically, and then continue to save in-place. this is an easy change (only the scroll operation has to be modified) and yet it adds a lot of reliability.

this highlights that saveAs()/saveAtomically() are not necessarily needed to solve (and are not the solution to) save()'s issues. saveAs()/saveAtomically(), if desired, can be added at a later stage.

(of course, in a perfect world you would detect scroll before starting save() and use saveAs() instead.)

**Comment by lazka** [comment]
> well, what can i say, my view is different.

Yeah, but I appreciate your input. I will ask some of the larger mutagen users and see what semantics they would expect or what API they want.

**Comment by catap** [comment]
@lazka any news about that?

**Comment by msikma** [comment]
Worth pointing out that this affects yt-dlp, which uses Mutagen to do metadata editing in some cases. Unfortunately, occasionally there's an issue where Mutagen fails at something, producing a corrupted file, which then causes the yt-dlp output file to be silently corrupted.

This is partly yt-dlp's fault, since yt-dlp weirdly assumes that Mutagen having a non-zero exit code means it's not installed (then it falls back on ffmpeg as alternative, but by then it's already working with a now corrupted file). But it's also a sensible assumption on yt-dlp's part that a program should either succeed as expected, or fail without corrupting the input, and that assumption is incorrect for Mutagen.

I would say that it's important for me to be able to assume that Mutagen will not corrupt the input file if it somehow isn't able to do its work for whatever reason—either it does its work, or it does nothing and returns non-zero. At the very least it would be great to have a command line flag for it, if it's not desirable to make it the new default.
